app/src/providers/evm/contracts/markets/contract.ts

```diff
@@ -0,0 +1,251 @@
+
+import MaketABI from "../../../../../../solidity/promptwars/artifacts/contracts/Market.sol/Market.json";
+
+type Timestamp = number;
+type WrappedBalance = number;
+
+type MarketData = {
+  imageUri: string;
+  startsAt: Timestamp;
+  endsAt: Timestamp;
+};
+
+type Player = {
+  id: string;
+  prompt: string;
+  outputImgUri: string;
+  result: string;
+  balance: number;
+};
+
+type CollateralToken = {
+  id: string;
+  balance: WrappedBalance;
+  decimals: number;
+  feeBalance: number;
+};
+
+type Resolution = {
+  window: number;
+  revealWindow: number;
+  resolvedAt: Timestamp;
+  playerId: string;
+};
+
+type Management = {
+  daoAccountId: string;
+  marketCreatorAccountId: string;
+  selfDestructWindow: Timestamp;
+  buySellThreshold: WrappedBalance;
+};
+
+type Fees = {
+  price: WrappedBalance;
+  feeRatio: WrappedBalance;
+  claimedAt: Timestamp;
+};
+
+export function walletClientToSigner(walletClient: WalletClient) {
+  const { account, chain, transport } = walletClient;
+  const network = {
+    chainId: chain.id,
+    name: chain.name,
+    ensAddress: chain.contracts?.ensRegistry?.address,
+  };
+  const provider = new BrowserProvider(transport, network);
+  const signer = new JsonRpcSigner(provider, account.address);
+
+  return signer;
+}
+
+/** Action to convert a viem Wallet Client to an ethers.js Signer. */
+export async function getEthersSigner({ chainId }: { chainId?: number } = {}) {
+  const walletClient = await getWalletClient({ chainId });
+
+  if (!walletClient) return undefined;
+
+  return walletClientToSigner(walletClient);
+}
+
+type PromptWarsMarketContractMethods = {
+  get_market_data: () => Promise<MarketData>;
+  get_resolution_data: () => Promise<Resolution>;
+  get_fee_data: () => Promise<Fees>;
+  get_management_data: () => Promise<Management>;
+  get_collateral_token_metadata: () => Promise<CollateralToken>;
+  get_block_timestamp: () => Promise<ethers.BigNumberish>;
+  resolved_at: () => Promise<Timestamp>;
+  get_amount_mintable: (amount: WrappedBalance) => Promise<[WrappedBalance, WrappedBalance]>;
+  get_amount_payable_unresolved: () => Promise<Array<WrappedBalance>>;
+  get_amount_payable_resolved: () => Promise<Array<WrappedBalance>>;
+  get_precision_decimals: () => Promise<WrappedBalance>;
+  // flags
+  is_resolved: () => Promise<boolean>;
+  is_open: () => Promise<boolean>;
+  is_over: () => Promise<boolean>;
+  is_reveal_window_expired: () => Promise<boolean>;
+  is_resolution_window_expired: () => Promise<boolean>;
+  is_self_destruct_window_expired: () => Promise<boolean>;
+  is_expired_unresolved: () => Promise<boolean>;
+  // mutators
+  sell: () => Promise<number>;
+  reveal: (playerId: string, result: string, outputImgUri: string) => Promise<void>;
+  resolve: (playerId: string) => Promise<boolean>;
+};
+
+class MarketContract {
+  private contract: ethers.Contract & PromptWarsMarketContractMethods;
+
+  private contractAddress: string;
+
```

Comment:
@netpoe I can create a class like this but as we are using wagmi. we can use their hooks instead of this and hooks support type inference when you add a const assertion to abi:

 https://wagmi.sh/react/typescript#contract-abis

Reply:
we would also have to use https://wagmi.sh/cli/getting-started

Reply:
I'll pull these changes and we'll collaborate on the best solution to this.